File: java/Rich_Game/test/com/tw/GameTest.java
package com.tw;

import com.tw.core.Game;
import com.tw.core.commands.CommandFactory;
import com.tw.core.map.*;
import com.tw.core.player.Player;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.*;
import static org.mockito.Mockito.mock;

/**
 * Created by pzzheng on 11/14/16.
 */
public class GameTest {
    @Test
    public void should_pass_turn_when_in_hospital_or_prison() {
        GameMap map = mock(GameMap.class);
        Game game = new Game(map);
        Player player = Player.createPlayerWith_Fund_Map_WAIT_TURN_STATE(map, 10000);
        Player player1 = Player.createPlayerWith_Fund_Map_WAIT_TURN_STATE(map, 10000);
        game.initPlayers(player, player1);

        assertThat(player.getStatus(), is(Player.Status.WAIT_FOR_COMMAND));
        assertThat(player1.getStatus(), is(Player.Status.WAIT_FOR_TURN));
        player1.stuckFor(2);

        game.nextPlayer();
        assertThat(player.getStatus(), is(Player.Status.WAIT_FOR_COMMAND));
        assertThat(player1.getStatus(), is(Player.Status.WAIT_FOR_TURN));


        game.nextPlayer();
        assertThat(player.getStatus(), is(Player.Status.WAIT_FOR_COMMAND));

        game.nextPlayer();
        assertThat(player1.getStatus(), is(Player.Status.WAIT_FOR_COMMAND));
    }

    @Test
    public void should_end_game_when_player_quit_game() {
        GameMap map = mock(GameMap.class);
        Game game = new Game(map);
        Player player = Player.createPlayerWith_Fund_Map_WAIT_TURN_STATE(map, 10000);
        Player player1 = Player.createPlayerWith_Fund_Map_WAIT_TURN_STATE(map, 10000);
        game.initPlayers(player, player1);

        assertThat(game.getStatus(), is(Game.Status.GAME_START));

        CommandFactory.Quit.execute(player1);

        assertThat(game.getStatus(), is(Game.Status.GAME_END));
    }

    @Test
    public void should_shift_to_next_player_when_player_end_turn() {
        GameMap map = mock(GameMap.class);
        Game game = new Game(map);
        Player player = Player.createPlayerWith_Fund_Map_WAIT_TURN_STATE(map, 10000);
        Player player1 = Player.createPlayerWith_Fund_Map_WAIT_TURN_STATE(map, 10000);
        game.initPlayers(player, player1);

        assertThat(player.getStatus(), is(Player.Status.WAIT_FOR_COMMAND));
        assertThat(player1.getStatus(), is(Player.Status.WAIT_FOR_TURN));

        player.endTurn();

        assertThat(player.getStatus(), is(Player.Status.WAIT_FOR_TURN));
        assertThat(player1.getStatus(), is(Player.Status.WAIT_FOR_COMMAND));
    }

    @Test
    public void should_end_game_when_only_one_player_not_bankrupt() {
        GameMap map = mock(GameMap.class);
        Game game = new Game(map);
        Player player = Player.createPlayerWith_Fund_Map_WAIT_TURN_STATE(map, 10000);
        Player player1 = Player.createPlayerWith_Fund_Map_WAIT_TURN_STATE(map, 10000);
        Player player2 = Player.createPlayerWith_Fund_Map_WAIT_TURN_STATE(map, 10000);
        game.initPlayers(player, player1, player2);

        player.bankrupt();
        player1.bankrupt();

        assertThat(game.getStatus(), is(Game.Status.GAME_END));
    }

}

File: ruby/rich_game_ruby/test/game_test.rb
require 'minitest/autorun'
require 'minitest/mock'
require '../src/game'
require '../src/places/starting'
require '../src/places/estate'
require '../src/player'
require '../src/game_map'
class GameTest < Minitest::Test

  def setup
    @starting = Starting.new
    @map = GameMap.new(2, 2, @starting, Estate.new(10), Estate.new(10), Estate.new(10))
  end

  def test_the_first_player_is_in_turn_and_the_starting_place_is_set
    game = Game.new(@map)
    player = Player.create_player_with_game_and_fund(game)
    player1 = Player.create_player_with_game_and_fund(game)
    game.initPlayers(player, player1)

    assert_equal game.currentPlayer, player
    assert_equal player.currentPlace, @starting
    assert_equal player1.currentPlace, @starting
    assert_equal player.status, Player::Status::WAIT_FOR_COMMAND
    assert_equal player1.status, Player::Status::WAIT_FOR_TURN
  end

  def test_that_can_let_the_current_player_execute_command
    game = Game.new(@map)
    player = Player.create_player_with_game_and_fund(game)
    player1 = Player.create_player_with_game_and_fund(game)
    game.initPlayers(player, player1)

    command = Minitest::Mock.new
    command.expect(:execute, Player::Status::WAIT_FOR_TURN, [Object])
    game.execute(command)

    assert_equal player.status, Player::Status::WAIT_FOR_TURN
  end

  def test_that_can_shift_to_next_player
    game = Game.new(@map)
    player = Player.create_player_with_game_and_fund(game)
    player1 = Player.create_player_with_game_and_fund(game)
    game.initPlayers(player, player1)

    game.nextPlayer()

    assert_equal game.currentPlayer, player1
    assert_equal player1.status, Player::Status::WAIT_FOR_COMMAND
  end

  def test_that_should_shift_to_next_player_if_currentPlayer_end_turn
    game = Game.new(@map)
    player = Player.create_player_with_game_and_fund(game)
    player1 = Player.create_player_with_game_and_fund(game)
    game.initPlayers(player, player1)

    player.endTurn

    assert_equal game.currentPlayer, player1
    assert_equal player1.status, Player::Status::WAIT_FOR_COMMAND
    assert_equal player.status, Player::Status::WAIT_FOR_TURN
  end

  def test_that_shift_to_player_after_next_player_if_next_player_bankrupt
    game = Game.new(@map)
    player = Player.create_player_with_game_and_fund(game)
    player1 = Player.create_player_with_game_and_fund(game)
    player2 = Player.create_player_with_game_and_fund(game)
    game.initPlayers(player, player1, player2)
    player1.bankrupt

    player.endTurn

    assert_equal game.currentPlayer, player2
    assert_equal player2.status, Player::Status::WAIT_FOR_COMMAND
    assert_equal player.status, Player::Status::WAIT_FOR_TURN
    assert_equal player1.status, Player::Status::BANKRUPT
  end

  def test_that_shift_to_player_after_next_player_if_next_player_in_hospital_or_prison
    game = Game.new(@map)
    player = Player.create_player_with_game_and_fund(game)
    player1 = Player.create_player_with_game_and_fund(game)
    player2 = Player.create_player_with_game_and_fund(game)
    game.initPlayers(player, player1, player2)
    player1.stuckFor(1)

    player.endTurn

    assert_equal game.currentPlayer, player2
    assert_equal player2.status, Player::Status::WAIT_FOR_COMMAND
    assert_equal player.status, Player::Status::WAIT_FOR_TURN
    assert_equal player1.status, Player::Status::WAIT_FOR_TURN
    refute player1.isStuck
  end

end